client/client.go

```diff
@@ -49,25 +53,46 @@ type Client struct {
 
 	ResponseTimeout time.Duration
 
-	ErrorHandler ErrorHandler
+	ErrorHandler    ErrorHandler
+	handleConnClose ConnCloseHandler
+	handleConnOpen  ConnOpenHandler
 }
 
 // Return a client.
-func New(network, addr string) (client *Client, err error) {
-	conn, err := net.Dial(network, addr)
+func NewNetClient(network, addr string) (client *Client, err error) {
 
-	if err != nil {
+	handlerConnOpen := func() (conn net.Conn, err error) {
+		log.Infof("Trying to connect to server %v ...", addr)
+		for num_tries := 1; ; num_tries++ {
+			if num_tries%100 == 0 {
+				log.Warningf("Still trying to connect to server %v, attempt# %v ...", addr, num_tries)
+			}
+			conn, err = net.Dial(network, addr)
+			if err != nil {
+				time.Sleep(500 * time.Millisecond)
```

Comment:
Why not just use `net.DialTimeout` ?

Reply:
Please a) Use an exponential backoff, b) Do not hard code the timeout values (a constructor like `New` can specify reasonable defaults, but as with the `New` `NewConnected` split that existed earlier you should allow overriding this value and probably also its unit without changing the client code.

Reply:
Since we're expecting the gearman to come back online within a short amount of time I didn't see that we want the exponential backoff because imagine that recovery took a few minutes and our exponential backoff would have already increased the retry delay to perhaps range of minutes, which would further delay coming back online, so the longer it takes to bring gearman online the worse our complete recovery is affected exponentially. I would say we use exponential backoff for resources that have unknown recovery goal (those that might never come back online?)

Reply:
Also `net.DialTimeout` does allow to specify the delay between tries, it allows to set your own timeout for various steps needed to establish the connection (e.g DNS lookup), Tested this without time.sleep() and this made hundreds of attempts to reconnect within seconds. 

> DialTimeout acts like Dial but takes a timeout.
> 
> The timeout includes name resolution, if required. When using TCP, and the host in the address parameter resolves to multiple IP addresses, the timeout is spread over each consecutive dial, such that each is given an appropriate fraction of the time to connect.